File: core/src/com/megaman/game/entities/impl/megaman/weapons/MegamanWeaponHandler.java
package com.megaman.game.entities.impl.megaman.weapons;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import com.megaman.game.ConstKeys;
import com.megaman.game.GameEngine;
import com.megaman.game.assets.SoundAsset;
import com.megaman.game.behaviors.BehaviorType;
import com.megaman.game.entities.utils.factories.EntityFactories;
import com.megaman.game.entities.EntityType;
import com.megaman.game.entities.utils.faceable.Facing;
import com.megaman.game.entities.impl.megaman.Megaman;
import com.megaman.game.entities.impl.projectiles.ChargeStatus;
import com.megaman.game.entities.impl.projectiles.Projectile;
import com.megaman.game.entities.impl.projectiles.ProjectileFactory;
import com.megaman.game.entities.impl.megaman.vals.MegamanVals;
import com.megaman.game.entities.impl.projectiles.impl.Bullet;
import com.megaman.game.entities.impl.projectiles.impl.ChargedShot;
import com.megaman.game.entities.impl.projectiles.impl.Fireball;
import com.megaman.game.utils.interfaces.Resettable;
import com.megaman.game.utils.interfaces.Updatable;
import com.megaman.game.utils.objs.Timer;
import com.megaman.game.world.BodySense;
import com.megaman.game.world.WorldVals;
import lombok.RequiredArgsConstructor;

import java.util.EnumMap;
import java.util.Iterator;
import java.util.Map;
import java.util.function.Supplier;

@RequiredArgsConstructor
public class MegamanWeaponHandler implements Updatable, Resettable {

    private static class MegaWeaponEntry implements Updatable {

        private final Timer cooldownTimer;

        private Supplier<Integer> normalCost;
        private Supplier<Integer> halfChargedCost;
        private Supplier<Integer> fullyChargedCost;
        private Supplier<Boolean> chargeable;
        private Supplier<Boolean> canFireWeapon;
        private Array<Projectile> spawned;
        private int ammo;

        private MegaWeaponEntry(float cooldownDur) {
            this(cooldownDur, () -> true);
        }

        private MegaWeaponEntry(float cooldownDur, Supplier<Boolean> chargeable) {
            this(cooldownDur, chargeable, () -> true);
        }

        private MegaWeaponEntry(float cooldownDur, Supplier<Boolean> chargeable, Supplier<Boolean> canFireWeapon) {
            this(cooldownDur, chargeable, canFireWeapon, () -> 0, () -> 0, () -> 0);
        }

        private MegaWeaponEntry(float cooldownDur, Supplier<Boolean> chargeable, Supplier<Boolean> canFireWeapon,
                                Supplier<Integer> normalCost, Supplier<Integer> halfChargedCost,
                                Supplier<Integer> fullyChargedCost) {
            this.canFireWeapon = canFireWeapon;
            this.chargeable = chargeable;
            this.normalCost = normalCost;
            this.halfChargedCost = halfChargedCost;
            this.fullyChargedCost = fullyChargedCost;
            cooldownTimer = new Timer(cooldownDur, true);
            spawned = new Array<>();
            ammo = MegamanVals.MAX_WEAPON_AMMO;
        }

        public int getNormalCost() {
            return normalCost.get();
        }

        public int getHalfChargedCost() {
            return halfChargedCost.get();
        }

        public int getFullyChargedCost() {
            return fullyChargedCost.get();
        }

        private boolean isChargeable() {
            return chargeable.get();
        }

        private boolean canFireWeapon() {
            return canFireWeapon.get();
        }

        @Override
        public void update(float delta) {
            cooldownTimer.update(delta);
            Iterator<Projectile> pIter = spawned.iterator();
            while (pIter.hasNext()) {
                Projectile p = pIter.next();
                if (p.dead) {
                    pIter.remove();
                }
            }
        }

    }

    private static final float MEGA_BUSTER_BULLET_VEL = 10f;
    private static final Vector2 FLAME_TOSS_TRAJ = new Vector2(35f, 10f);

    private final Megaman megaman;
    private final GameEngine engine;
    private final EntityFactories factories;
    private final Map<MegamanWeapon, MegaWeaponEntry> weapons;

    public MegamanWeaponHandler(Megaman megaman) {
        this.megaman = megaman;
        engine = megaman.game.getGameEngine();
        factories = megaman.game.getEntityFactories();
        weapons = new EnumMap<>(MegamanWeapon.class);
    }

    @Override
    public void reset() {
        for (MegaWeaponEntry e : weapons.values()) {
            e.cooldownTimer.setToEnd();
        }
    }

    @Override
    public void update(float delta) {
        MegaWeaponEntry e = weapons.get(megaman.currWeapon);
        if (e == null) {
            return;
        }
        e.update(delta);
    }

    private MegaWeaponEntry getWeaponEntry(MegamanWeapon weapon) {
        return switch (weapon) {
            case MEGA_BUSTER -> new MegaWeaponEntry(.01f);
            case FLAME_TOSS -> {
                MegaWeaponEntry e = new MegaWeaponEntry(.5f);
                e.normalCost = () -> 3;
                e.halfChargedCost = () -> 5;
                e.fullyChargedCost = () -> 7;
                e.chargeable = () -> !megaman.is(BodySense.BODY_IN_WATER);
                e.canFireWeapon = () -> !megaman.is(BodySense.BODY_IN_WATER) && e.spawned.size == 0;
                yield e;
            }
        };
    }

    public Array<Projectile> getSpawned(MegamanWeapon weapon) {
        if (!hasWeapon(weapon)) {
            throw new IllegalStateException("Megaman does not have the weapon: " + weapon);
        }
        return weapons.get(weapon).spawned;
    }

    public void putWeapon(MegamanWeapon weapon) {
        weapons.put(weapon, getWeaponEntry(weapon));
    }

    public boolean hasWeapon(MegamanWeapon weapon) {
        return weapons.containsKey(weapon);
    }

    public boolean canFireWeapon(MegamanWeapon weapon, ChargeStatus stat) {
        if (!hasWeapon(weapon)) {
            return false;
        }
        MegaWeaponEntry e = weapons.get(weapon);
        if (!e.cooldownTimer.isFinished() || !e.canFireWeapon()) {
            return false;
        }
        int cost = e.isChargeable() ?
                (weapon == MegamanWeapon.MEGA_BUSTER ? 0 : switch (stat) {
                    case FULLY_CHARGED -> e.getFullyChargedCost();
                    case HALF_CHARGED -> e.getHalfChargedCost();
                    case NOT_CHARGED -> e.getNormalCost();
                }) : e.getNormalCost();
        return cost <= e.ammo;
    }

    public boolean isChargeable(MegamanWeapon weapon) {
        return hasWeapon(weapon) && weapons.get(weapon).isChargeable();
    }

    public void translateAmmo(MegamanWeapon weapon, int delta) {
        if (!hasWeapon(weapon)) {
            throw new IllegalStateException("Megaman does not have the weapon: " + weapon);
        }
        MegaWeaponEntry e = weapons.get(weapon);
        e.ammo += delta;
        if (e.ammo >= MegamanVals.MAX_WEAPON_AMMO) {
            e.ammo = MegamanVals.MAX_WEAPON_AMMO;
        } else if (e.ammo < 0) {
            e.ammo = 0;
        }
    }

    public void setToMaxAmmo(MegamanWeapon weapon) {
        if (!hasWeapon(weapon)) {
            throw new IllegalStateException("Megaman does not have the weapon: " + weapon);
        }
        MegaWeaponEntry e = weapons.get(weapon);
        e.ammo = MegamanVals.MAX_WEAPON_AMMO;
    }

    public void depleteAmmo(MegamanWeapon weapon) {
        if (!hasWeapon(weapon)) {
            throw new IllegalStateException("Megaman does not have the weapon: " + weapon);
        }
        MegaWeaponEntry e = weapons.get(weapon);
        e.ammo = 0;
    }

    public int getAmmo(MegamanWeapon weapon) {
        if (weapon == MegamanWeapon.MEGA_BUSTER) {
            return Integer.MAX_VALUE;
        }
        return weapons.get(weapon).ammo;
    }

    public boolean fireWeapon(MegamanWeapon weapon, ChargeStatus stat) {
        if (!canFireWeapon(weapon, stat)) {
            return false;
        }
        if (!isChargeable(weapon)) {
            stat = ChargeStatus.NOT_CHARGED;
        }
        MegaWeaponEntry e = weapons.get(weapon);
        int cost = weapon == MegamanWeapon.MEGA_BUSTER ? 0 : switch (stat) {
            case FULLY_CHARGED -> e.getFullyChargedCost();
            case HALF_CHARGED -> e.getHalfChargedCost();
            case NOT_CHARGED -> e.getNormalCost();
        };
        if (cost > getAmmo(weapon)) {
            return false;
        }
        Projectile p = switch (weapon) {
            case MEGA_BUSTER -> fireMegaBuster(stat);
            case FLAME_TOSS -> fireFlameToss(stat);
        };
        if (p != null) {
            e.spawned.add(p);
        }
        e.cooldownTimer.reset();
        translateAmmo(weapon, -cost);
        return true;
    }

    private Vector2 getSpawnCenter() {
        Vector2 spawnCenter = new Vector2(megaman.body.getCenter());
        float xOffset = WorldVals.PPM * .85f;
        if (megaman.is(Facing.LEFT)) {
            xOffset *= -1f;
        }
        spawnCenter.x += xOffset;
        float yOffset = WorldVals.PPM / 16f;
        if (megaman.is(BehaviorType.CLIMBING) || megaman.is(BehaviorType.WALL_SLIDING)) {
            yOffset += .15f * WorldVals.PPM;
        } else if (megaman.is(BodySense.FEET_ON_GROUND)) {
            yOffset -= .05f * WorldVals.PPM;
        } else {
            yOffset += .25f * WorldVals.PPM;
        }
        spawnCenter.y += yOffset;
        return spawnCenter;
    }

    private Projectile fireMegaBuster(ChargeStatus stat) {
        float x = MEGA_BUSTER_BULLET_VEL;
        if (megaman.is(Facing.LEFT)) {
            x *= -1f;
        }
        Vector2 trajectory = new Vector2(x, 0f);
        ObjectMap<String, Object> data = new ObjectMap<>();
        data.put(ConstKeys.OWNER, megaman);
        data.put(ConstKeys.TRAJECTORY, trajectory);
        Projectile proj = switch (stat) {
            case NOT_CHARGED -> (Bullet) factories.fetch(EntityType.PROJECTILE, ProjectileFactory.BULLET);
            case HALF_CHARGED, FULLY_CHARGED -> {
                data.put(ConstKeys.BOOL, stat == ChargeStatus.FULLY_CHARGED);
                yield (ChargedShot) factories.fetch(EntityType.PROJECTILE, ProjectileFactory.CHARGED_SHOT);
            }
        };
        if (stat == ChargeStatus.NOT_CHARGED) {
            megaman.request(SoundAsset.MEGA_BUSTER_BULLET_SHOT_SOUND, true);
        } else {
            megaman.request(SoundAsset.MEGA_BUSTER_CHARGED_SHOT_SOUND, true);
            megaman.request(SoundAsset.MEGA_BUSTER_CHARGING_SOUND, false);
        }
        Vector2 s = getSpawnCenter();
        if (megaman.isUpsideDown()) {
            if (megaman.is(BehaviorType.CLIMBING)) {
                s.y -= .45f * WorldVals.PPM;
            } else {
                s.y -= .05f * WorldVals.PPM;
            }
        }
        engine.spawn(proj, s, data);
        return proj;
    }

    private Projectile fireFlameToss(ChargeStatus stat) {
        ObjectMap<String, Object> data = new ObjectMap<>();
        data.put(ConstKeys.OWNER, megaman);
        Projectile proj = switch (stat) {
            case NOT_CHARGED, HALF_CHARGED, FULLY_CHARGED -> {
                data.put(ConstKeys.LEFT, megaman.is(Facing.LEFT));
                yield (Fireball) factories.fetch(EntityType.PROJECTILE, ProjectileFactory.FIREBALL);
            }

            // TODO: charged flame weapons
            /*
            case HALF_CHARGED -> {

            }
            case FULLY_CHARGED -> {

            }
             */

        };
        engine.spawn(proj, getSpawnCenter(), data);

        // TODO: playMusic different sounds
        megaman.request(SoundAsset.CRASH_BOMBER_SOUND, true);

        return proj;
    }

}

File: core/src/main/kotlin/entities/megaman/MegamanWeaponHandler.kt
package com.megaman.maverick.game.entities.megaman

import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ObjectMap
import com.mega.game.engine.common.enums.Direction
import com.mega.game.engine.common.extensions.equalsAny
import com.mega.game.engine.common.interfaces.Resettable
import com.mega.game.engine.common.interfaces.Updatable
import com.mega.game.engine.common.objects.Properties
import com.mega.game.engine.common.time.Timer
import com.megaman.maverick.game.ConstKeys
import com.megaman.maverick.game.ConstVals
import com.megaman.maverick.game.assets.SoundAsset
import com.megaman.maverick.game.behaviors.BehaviorType
import com.megaman.maverick.game.entities.EntityType
import com.megaman.maverick.game.entities.contracts.AbstractProjectile
import com.megaman.maverick.game.entities.factories.EntityFactories
import com.megaman.maverick.game.entities.factories.impl.ProjectilesFactory
import com.megaman.maverick.game.entities.megaman.components.GROUND_SLIDE_SPRITE_OFFSET_Y
import com.megaman.maverick.game.entities.megaman.constants.MegaChargeStatus
import com.megaman.maverick.game.entities.megaman.constants.MegamanValues
import com.megaman.maverick.game.entities.megaman.constants.MegamanWeapon
import com.megaman.maverick.game.entities.projectiles.Fireball
import com.megaman.maverick.game.world.body.BodySense
import com.megaman.maverick.game.world.body.getCenter
import com.megaman.maverick.game.world.body.isSensing

class MegaWeaponEntry(
    cooldownDur: Float,
    var chargeable: () -> Boolean = { true },
    var canFireWeapon: () -> Boolean = { true },
    var normalCost: () -> Int = { 0 },
    var halfChargedCost: () -> Int = { 0 },
    var fullyChargedCost: () -> Int = { 0 },
    var cooldownTimer: Timer = Timer(cooldownDur).setToEnd(),
    var spawned: Array<AbstractProjectile> = Array(),
    var ammo: Int = MegamanValues.MAX_WEAPON_AMMO
) : Updatable {

    override fun update(delta: Float) {
        cooldownTimer.update(delta)
        spawned.removeAll { it.dead }
    }
}

class MegamanWeaponHandler(private val megaman: Megaman) : Updatable, Resettable {

    companion object {
        private const val MEGA_BUSTER_BULLET_VEL = 10f
    }

    private val engine = megaman.game.engine
    private val weapons = ObjectMap<MegamanWeapon, MegaWeaponEntry>()
    private val spawnCenter: Vector2
        get() {
            val spawnCenter = Vector2(megaman.body.getCenter())

            val xOffset = ConstVals.PPM * megaman.facing.value *
                if (megaman.isBehaviorActive(BehaviorType.RIDING_CART)) {
                    if (megaman.body.isSensing(BodySense.FEET_ON_GROUND)) 1.25f else 1f
                } else if (!megaman.body.isSensing(BodySense.FEET_ON_GROUND)) 0.75f
                else if (megaman.isBehaviorActive(BehaviorType.GROUND_SLIDING)) 0.25f
                else if (megaman.slipSliding) 0.65f
                else 1f

            var yOffset = ConstVals.PPM / 16f

            /*
            if (!megaman.body.isSensing(BodySense.FEET_ON_GROUND) &&
                !megaman.isAnyBehaviorActive(BehaviorType.CLIMBING, BehaviorType.WALL_SLIDING, BehaviorType.RIDING_CART)
            ) yOffset += 0.15f * ConstVals.PPM
             */

            yOffset += if (megaman.isAnyBehaviorActive(BehaviorType.JETPACKING, BehaviorType.WALL_SLIDING))
                0.3f * ConstVals.PPM
            else if (megaman.isBehaviorActive(BehaviorType.GROUND_SLIDING)) -0.05f * ConstVals.PPM
            else if (megaman.isBehaviorActive(BehaviorType.RIDING_CART)) {
                if (megaman.body.isSensing(BodySense.FEET_ON_GROUND)) 0.6f * ConstVals.PPM else 0.3f * ConstVals.PPM
            } // else if (megaman.isBehaviorActive(BehaviorType.WALL_SLIDING)) 0.4f * ConstVals.PPM
            else if (megaman.direction == Direction.DOWN) {
                (if (megaman.body.isSensing(BodySense.FEET_ON_GROUND)) 0.15f else 0.25f) * ConstVals.PPM
            } else 0.135f * ConstVals.PPM
            /*
            else if (megaman.isBehaviorActive(BehaviorType.CLIMBING)) 0.15f * ConstVals.PPM
            else if (megaman.body.isSensing(BodySense.FEET_ON_GROUND)) 0.15f * ConstVals.PPM
            else 0.15f * ConstVals.PPM
             */

            if (megaman.direction.isVertical()) {
                spawnCenter.x += xOffset
                spawnCenter.y += if (megaman.direction == Direction.DOWN) (-yOffset + 0.1f * ConstVals.PPM) else yOffset

                if (megaman.isBehaviorActive(BehaviorType.GROUND_SLIDING)) {
                    var groundSlideOffset = GROUND_SLIDE_SPRITE_OFFSET_Y * ConstVals.PPM
                    if (megaman.direction == Direction.UP) spawnCenter.y -= groundSlideOffset
                    else spawnCenter.y += groundSlideOffset
                }
            } else {
                yOffset += if (megaman.body.isSensing(BodySense.FEET_ON_GROUND)) -0.075f * ConstVals.PPM
                else if (megaman.isBehaviorActive(BehaviorType.WALL_SLIDING)) -0.1f * ConstVals.PPM
                else 0.05f * ConstVals.PPM
                spawnCenter.x += if (megaman.direction == Direction.LEFT) -yOffset else yOffset
                spawnCenter.y += xOffset
            }

            return spawnCenter
        }

    override fun reset() = weapons.values().forEach { it.cooldownTimer.setToEnd() }

    override fun update(delta: Float) {
        weapons[megaman.currentWeapon]?.update(delta)
    }

    fun getSpawned(weapon: MegamanWeapon) = weapons[weapon]?.spawned

    fun putWeapon(weapon: MegamanWeapon): MegaWeaponEntry? = weapons.put(weapon, getWeaponEntry(weapon))

    fun hasWeapon(weapon: MegamanWeapon) = weapons.containsKey(weapon)

    fun canFireWeapon(weapon: MegamanWeapon, stat: MegaChargeStatus): Boolean {
        if (!megaman.ready || !hasWeapon(weapon)) return false

        val e = weapons[weapon]
        if (!e.cooldownTimer.isFinished() || !e.canFireWeapon()) return false

        val cost = if (e.chargeable()) (if (weapon === MegamanWeapon.BUSTER) 0
        else when (stat) {
            MegaChargeStatus.FULLY_CHARGED -> e.fullyChargedCost()
            MegaChargeStatus.HALF_CHARGED -> e.halfChargedCost()
            MegaChargeStatus.NOT_CHARGED -> e.normalCost()
        })
        else e.normalCost()

        return cost <= e.ammo
    }

    fun isChargeable(weapon: MegamanWeapon) = hasWeapon(weapon) && weapons[weapon].chargeable()

    fun translateAmmo(weapon: MegamanWeapon, delta: Int) {
        if (!hasWeapon(weapon)) return

        val weaponEntry = weapons[weapon]
        weaponEntry.ammo += delta

        if (weaponEntry.ammo >= MegamanValues.MAX_WEAPON_AMMO) weaponEntry.ammo = MegamanValues.MAX_WEAPON_AMMO
        else if (weaponEntry.ammo < 0) weaponEntry.ammo = 0
    }

    fun setAllToMaxAmmo() = weapons.keys().forEach { setToMaxAmmo(it) }

    fun setToMaxAmmo(weapon: MegamanWeapon) {
        weapons[weapon]?.ammo = MegamanValues.MAX_WEAPON_AMMO
    }

    fun depleteAmmo(weapon: MegamanWeapon) {
        weapons[weapon]?.ammo = 0
    }

    fun getAmmo(weapon: MegamanWeapon) = if (!hasWeapon(weapon)) 0
    else if (weapon == MegamanWeapon.BUSTER) Int.MAX_VALUE else weapons[weapon].ammo

    fun isDepleted(weapon: MegamanWeapon) = getAmmo(weapon) == 0

    fun fireWeapon(weapon: MegamanWeapon, stat: MegaChargeStatus): Boolean {
        var _stat: MegaChargeStatus = stat
        if (!canFireWeapon(weapon, _stat)) return false
        if (!isChargeable(weapon)) _stat = MegaChargeStatus.NOT_CHARGED

        val weaponEntry = weapons[weapon]
        val cost = if (weapon.equalsAny(MegamanWeapon.BUSTER, MegamanWeapon.RUSH_JETPACK)) 0
        else when (_stat) {
            MegaChargeStatus.FULLY_CHARGED -> weaponEntry.fullyChargedCost()
            MegaChargeStatus.HALF_CHARGED -> weaponEntry.halfChargedCost()
            MegaChargeStatus.NOT_CHARGED -> weaponEntry.normalCost()
        }
        if (cost > getAmmo(weapon)) return false

        val projectile = when (weapon) {
            MegamanWeapon.BUSTER, MegamanWeapon.RUSH_JETPACK -> fireMegaBuster(_stat)
            // TODO: MegamanWeapon.FLAME_TOSS -> fireFlameToss(_stat)
        }

        weaponEntry.spawned.add(projectile)
        weaponEntry.cooldownTimer.reset()
        translateAmmo(weapon, -cost)

        return true
    }

    private fun getWeaponEntry(weapon: MegamanWeapon) = when (weapon) {
        MegamanWeapon.BUSTER -> MegaWeaponEntry(.01f)
        MegamanWeapon.RUSH_JETPACK -> MegaWeaponEntry(cooldownDur = 0.1f, chargeable = { false })
        /* TODO:
        MegamanWeapon.FLAME_TOSS -> {
            val e = MegaWeaponEntry(.5f)
            e.normalCost = { 3 }
            e.halfChargedCost = { 5 }
            e.fullyChargedCost = { 7 }
            e.chargeable = { !megaman.body.isSensing(BodySense.IN_WATER) }
            e.canFireWeapon = { !megaman.body.isSensing(BodySense.IN_WATER) && e.spawned.size == 0 }
            e
        }
         */
    }

    private fun fireMegaBuster(stat: MegaChargeStatus): AbstractProjectile {
        val trajectory = Vector2()
        if (megaman.direction.isVertical()) trajectory.x = MEGA_BUSTER_BULLET_VEL * megaman.facing.value
        else trajectory.y = MEGA_BUSTER_BULLET_VEL * megaman.facing.value
        trajectory.scl(ConstVals.PPM.toFloat())

        if (megaman.applyMovementScalarToBullet) trajectory.scl(megaman.movementScalar)

        val props = Properties()
        props.put(ConstKeys.OWNER, megaman)
        props.put(ConstKeys.TRAJECTORY, trajectory)
        props.put(ConstKeys.DIRECTION, megaman.direction)

        val megaBusterShot = when (stat) {
            MegaChargeStatus.NOT_CHARGED -> EntityFactories.fetch(EntityType.PROJECTILE, ProjectilesFactory.BULLET)

            MegaChargeStatus.HALF_CHARGED, MegaChargeStatus.FULLY_CHARGED -> {
                props.put(ConstKeys.BOOLEAN, stat == MegaChargeStatus.FULLY_CHARGED)
                EntityFactories.fetch(EntityType.PROJECTILE, ProjectilesFactory.CHARGED_SHOT)
            }
        } ?: throw IllegalStateException("MegaBusterShot is null")

        if (stat === MegaChargeStatus.NOT_CHARGED) megaman.requestToPlaySound(
            SoundAsset.MEGA_BUSTER_BULLET_SHOT_SOUND, false
        )
        else {
            megaman.requestToPlaySound(SoundAsset.MEGA_BUSTER_CHARGED_SHOT_SOUND, false)
            megaman.stopSound(SoundAsset.MEGA_BUSTER_CHARGING_SOUND)
        }

        val s = spawnCenter
        props.put(ConstKeys.POSITION, s)
        engine.spawn(megaBusterShot, props)

        return megaBusterShot as AbstractProjectile
    }

    private fun fireFlameToss(stat: MegaChargeStatus): AbstractProjectile {
        val props = Properties()
        props.put(ConstKeys.OWNER, megaman)
        val fireball = when (stat) {
            MegaChargeStatus.NOT_CHARGED, MegaChargeStatus.HALF_CHARGED, MegaChargeStatus.FULLY_CHARGED -> {
                props.put(
                    ConstKeys.TRAJECTORY, Vector2(
                        MegamanValues.FLAME_TOSS_X_VEL * megaman.facing.value, MegamanValues.FLAME_TOSS_Y_VEL
                    ).scl(ConstVals.PPM.toFloat())
                )
                props.put(ConstKeys.GRAVITY, Vector2(0f, MegamanValues.FLAME_TOSS_GRAVITY))
                props.put(Fireball.BURST_ON_HIT_BODY, true)
                props.put(Fireball.BURST_ON_DAMAGE_INFLICTED, true)
                EntityFactories.fetch(EntityType.PROJECTILE, ProjectilesFactory.FIREBALL) as Fireball
            }
        }
        props.put(ConstKeys.POSITION, spawnCenter)
        engine.spawn(fireball, props)
        megaman.requestToPlaySound(SoundAsset.CRASH_BOMBER_SOUND, false)
        return fireball
    }
}
